core/arch/arm/mm/core_mmu.c

```diff
@@ -2287,3 +2286,68 @@ vaddr_t io_pa_or_va_nsec(struct io_pa_va *p)
 	}
 	return p->pa;
 }
+
+#ifdef CFG_CORE_RESERVED_SHM
+static TEE_Result teecore_init_pub_ram(void)
+{
+	vaddr_t s = 0;
+	vaddr_t e = 0;
+
+	/* get virtual addr/size of NSec shared mem allcated from teecore */
+	core_mmu_get_mem_by_type(MEM_AREA_NSEC_SHM, &s, &e);
+
+	if (s >= e || s & SMALL_PAGE_MASK || e & SMALL_PAGE_MASK)
+		panic("invalid PUB RAM");
+
+	/* extra check: we could rely on  core_mmu_get_mem_by_type() */
+	if (!tee_vbuf_is_non_sec(s, e - s))
+		panic("PUB RAM is not non-secure");
+
+#ifdef CFG_PL310
+	/* Allocate statically the l2cc mutex */
+	tee_l2cc_store_mutex_boot_pa(virt_to_phys((void *)s));
+	s += sizeof(uint32_t);			/* size of a pl310 mutex */
+	s =  ROUNDUP(s, SMALL_PAGE_SIZE);	/* keep required alignment */
+#endif
+
+	default_nsec_shm_paddr = virt_to_phys((void *)s);
+	default_nsec_shm_size = e - s;
+
+	return TEE_SUCCESS;
+}
+early_init(teecore_init_pub_ram);
+#endif /*CFG_CORE_RESERVED_SHM*/
+
+void core_mmu_init_ta_ram(void)
+{
+	vaddr_t s = 0;
+	vaddr_t e = 0;
+	paddr_t ps = 0;
+	paddr_t pe = 0;
+
+	/* get virtual addr/size of RAM where TA are loaded/executedNSec
+	 * shared mem allcated from teecore */
+#ifndef CFG_VIRTUALIZATION
+	core_mmu_get_mem_by_type(MEM_AREA_TA_RAM, &s, &e);
+#else
+	virt_get_ta_ram(&s, &e);
+#endif
+	ps = virt_to_phys((void *)s);
+	pe = virt_to_phys((void *)(e - 1)) + 1;
+
+	if (!ps || (ps & CORE_MMU_USER_CODE_MASK) ||
+	    !pe || (pe & CORE_MMU_USER_CODE_MASK))
+		panic("invalid TA RAM");
+
+	/* extra check: we could rely on  core_mmu_get_mem_by_type() */
+	if (!tee_pbuf_is_sec(ps, pe - ps))
+		panic("TA RAM is not secure");
+
+	if (!tee_mm_is_empty(&tee_mm_sec_ddr))
+		panic("TA RAM pool is not empty");
```

Comment:
should be fixed:
```patch
-		panic("TA RAM pool is not empty");
+		panic("TA RAM pool is empty");
```

Reply:
That would be incorrect.

Reply:
Sorry, you're right. Actually the error message looks strange. It's the allocation reference list  in the pool that is not empty (and should be), not the pool itself. But that's not related to this change.
Discard my comment.